File: Quick Sort/quick-sort.py
import random

class QuickSort:
    def __init__(self) -> None:
        pass

    def partition(self, arr, low, high):
        pivot = arr[low]
        i = low + 1
        j = high

        while True:
            while i <= high and arr[i] <= pivot:
                i += 1

            while j > low and arr[j] > pivot:
                j -= 1

            if i >= j:
                break

            arr[i], arr[j] = arr[j], arr[i]

        arr[low], arr[j] = arr[j], arr[low]
        return j

    def deterministic_quick_sort(self, arr, low, high):
        if low < high:
            p = self.partition(arr, low, high)
            self.deterministic_quick_sort(arr, low, p)
            self.deterministic_quick_sort(arr, p + 1, high)

    def random_quick_sort(self, arr, low, high):
        if low < high:
            r = random.randint(low, high)
            arr[low], arr[r] = arr[r], arr[low]
            p = self.partition(arr, low, high)
            self.random_quick_sort(arr, low, p)
            self.random_quick_sort(arr, p + 1, high)


l1 = [29, 4, 11, 33, -1, 0, 55, 24]
l2 = [29, 4, 11, 33, -1, 0, 55, 24]
qs = QuickSort()
qs.deterministic_quick_sort(l1, 0, len(l1) - 1)
print(l1)           
qs.random_quick_sort(l2, 0, len(l2) - 1)
print(l2)                    

File: Quick Sort/quick_sort.cpp
#include<bits/stdc++.h>
using namespace std;

class QuickSort {
    int partition(vector<int> &arr, int low, int high) {
        int pivot = arr[low]; //pivot as first element
        int i = low + 1;
        int j = high;

        while (true) {
            while (i <= high && arr[i] <= pivot) {
                i++;
            }

            while (j > low && arr[j] > pivot) {
                j--;
            }

            if (i >= j) {
                break;
            }

            swap(arr[i], arr[j]);
        }

        swap(arr[low], arr[j]);
        return j;
    }

public:
    void deterministicQuickSort(vector<int> &arr, int low, int high) {
        if(low < high) {
            int pivot = partition(arr, low, high);
            deterministicQuickSort(arr, low, pivot);
            deterministicQuickSort(arr, pivot + 1, high);
        }
    }

    void randomQuickSort(vector<int> &arr, int low, int high) {
        if(low < high) {
            int r = low + rand() % (high - low + 1);
            swap(arr[r], arr[low]);
            int pivot = partition(arr, low, high);
            randomQuickSort(arr, low, pivot);
            randomQuickSort(arr, pivot + 1, high);
        }
    }
};


int main() {
    srand(time(0));
    QuickSort qs;

    vector<int> x = {15, 10, 12, 20, 1, 4, 17, 16, 11};
    int n = x.size();

    qs.deterministicQuickSort(x, 0, n-1);
    for (auto i : x)
        cout << i << " ";
    cout << endl;

    x = {5, 4, 1, 2, 3};
    n = x.size();
    qs.randomQuickSort(x, 0, n-1);
    for (auto i : x)
        cout << i << " ";
    cout << endl;

    return 0;
}